library/core/src/intrinsics.rs

```diff
@@ -1207,29 +1207,32 @@ extern "rust-intrinsic" {
 
     /// Reinterprets the bits of a value of one type as another type.
     ///
-    /// Both types must have the same size. Neither the original, nor the result,
-    /// may be an [invalid value](../../nomicon/what-unsafe-does.html).
+    /// Both types must have the same size. Compilation will fail if this is not guaranteed.
     ///
     /// `transmute` is semantically equivalent to a bitwise move of one type
     /// into another. It copies the bits from the source value into the
-    /// destination value, then forgets the original. It's equivalent to C's
-    /// `memcpy` under the hood, just like `transmute_copy`.
+    /// destination value, then forgets the original. Note that source and destination
+    /// are passed by-value, which means if `T` or `U` contains padding, that padding
+    /// might *not* be preserved by `transmute`.
+    ///
+    /// Both the argument and the result must be [valid](../../nomicon/what-unsafe-does.html) at
+    /// their given type. Violating this condition leads to [undefined behavior][ub]. The compiler
+    /// will generate code *assuming that you, the programmer, ensure that there will never be
+    /// undefined behavior*. It is therefore your responsibility to guarantee that every value
+    /// passed to `transmute` is valid at both types `T` and `U`. Failing to uphold this condition
```

Comment:
I think it's a little redundant to say that `T` must be valid too, since you already have the `T` so if it's invalid, the jig's already up -- you're hosed whether or not you pass it to transmute.

That said, being overly clear here doesn't hurt, and this was present in the previous version of the documentation as well.

Reply:
I think the text flows better with the implied symmetry of "your value has to be a value that is valid if interpreted in either direction". It also helps to have the explicit clarity that, yes, if you create a `T` with an invalid value (somehow), the transmute can't "fix" the invalidity of the value by making it a `U`, even if that value is valid at `U`: it's invalid at `T`! This conveys e.g. the semantics of the value getting tagged as poison or whatever and the fact that UB has already "escaped containment" and is now free to roam in the program.

The only doors it closes off are ones **no sane program** should arrive at, yes, even amongst `unsafe` programs, but also we should assume anyone wanting to use `mem::transmute` has good reasons for gazing into the face of madness, and thus has abandoned "common sense".

Reply:
@workingjubilee has summarized my reasoning perfectly. :)